File: src/detfm/simplify.cpp
#include "detfm/simplify.hpp"
#include "detfm/common.hpp"
#include "detfm/opinfo.hpp"
#include <abc/AbcFile.hpp>
#include <abc/parser/Parser.hpp>
#include <cmath>
#include <functional>
#include <memory>
#include <optional>
#include <stack>
#include <stdexcept>
#include <unordered_map>
#include <variant>
#include <vector>

namespace abc = swf::abc;
using namespace abc::parser;

using StackValue = std::variant<std::monostate, bool, double, std::string>;
using Operation  = std::function<StackValue(const StackValue&, const StackValue&)>;

std::vector<std::unordered_map<OP, Operation>> operations
    = { {}, // monostate
        {}, // bool
        {
            // double
            { OP::add, [](auto& a, auto& b) { return std::get<2>(b) + std::get<2>(a); } },
            { OP::subtract, [](auto& a, auto& b) { return std::get<2>(b) - std::get<2>(a); } },
            { OP::multiply, [](auto& a, auto& b) { return std::get<2>(b) * std::get<2>(a); } },
            { OP::divide, [](auto& a, auto& b) { return std::get<2>(b) / std::get<2>(a); } },
        },
        {
            // string
            { OP::add, [](auto& a, auto& b) { return std::get<3>(b) + std::get<3>(a); } },
        } };

std::unordered_map<OP, std::array<uint32_t, 2>> stack_operations = {
    // { opcode, {take from stack, put on stack}}
    // stack indenpendent
    { OP::jump, {} },
    { OP::kill, {} },
    { OP::popscope, {} },
    { OP::returnvoid, {} },
    { OP::returnvoid, {} },

    // add to the stack
    { OP::getlocal0, { 0, 1 } },
    { OP::getlex, { 0, 1 } },
    { OP::pushnull, { 0, 1 } },
    { OP::findproperty, { 0, 1 } },
    { OP::findpropstrict, { 0, 1 } },
    { OP::newarray, { 0, 1 } },
    { OP::newfunction, { 0, 1 } },
    { OP::newcatch, { 0, 2 } }, // should be 1, but also add the error to the stack

    // take from the stack
    { OP::setlocal1, { 1, 0 } },
    { OP::pushscope, { 1, 0 } },
    { OP::pop, { 1, 0 } },
    { OP::callpropvoid, { 1, 0 } },
    { OP::iffalse, { 1, 0 } },
    { OP::iftrue, { 1, 0 } },
    { OP::ifeq, { 2, 0 } },
    { OP::setproperty, { 2, 0 } },
    { OP::initproperty, { 2, 0 } },
    { OP::setslot, { 2, 0 } },

    // hybrid
    { OP::getproperty, { 1, 1 } },
    { OP::constructprop, { 1, 1 } },
    { OP::construct, { 1, 1 } },
    { OP::applytype, { 1, 1 } },
    { OP::callproperty, { 1, 1 } },
    { OP::equals, { 2, 1 } },
};

std::unordered_map<OP, uint32_t> dynamic_stack_op = {
    // ins->args[0]
    { OP::construct, 0 },
    { OP::applytype, 0 },
    { OP::newarray, 0 },
    // ins->args[1]
    { OP::constructprop, 1 },
    { OP::callproperty, 1 },
    { OP::callpropvoid, 1 },
};

template <typename T> T pop_value(std::stack<T>& stack) {
    auto v = std::move(stack.top());
    stack.pop();
    return v;
}

void edit_ins(
    std::shared_ptr<abc::AbcFile>& abc, Parser& parser, OpRegister& insreg,
    std::stack<StackValue> stack, std::shared_ptr<OpInfo>& opinfo, uint32_t ins2remove) {
    for (uint32_t i = 0; i < ins2remove; ++i)
        insreg[opinfo->ins->prev.lock()->addr]->remove(parser, insreg);

    if (std::holds_alternative<double>(stack.top())) {
        const auto& value = std::get<double>(stack.top());
        if (std::fmod(value, 1) != 0 || std::abs(value) > 0x8000) {
            uint32_t index = abc->cpool.doubles.size();
            abc->cpool.doubles.push_back(value);
            opinfo->ins->opcode = OP::pushdouble;
            opinfo->ins->args   = { index };
        } else {
            opinfo->ins->opcode = std::abs(value) > 0x80 ? OP::pushshort : OP::pushbyte;
            opinfo->ins->args   = { static_cast<uint32_t>(value) };
        }
    } else if (std::holds_alternative<std::string>(stack.top())) {
        uint32_t index = abc->cpool.strings.size();
        abc->cpool.strings.push_back(std::get<std::string>(stack.top()));
        opinfo->ins->opcode = OP::pushstring;
        opinfo->ins->args   = { index };
    } else if (std::holds_alternative<bool>(stack.top())) {
        opinfo->ins->opcode = std::get<bool>(stack.top()) ? OP::pushtrue : OP::pushfalse;
        opinfo->ins->args   = {};
    }
}

void simplify_expressions(std::shared_ptr<abc::AbcFile>& abc, abc::Method& method) {
    Parser parser(method);
    std::stack<StackValue> stack;

    OpRegister insreg;
    // populate the instruction register
    auto ins  = parser.begin;
    auto prev = insreg[ins->addr] = std::make_shared<OpInfo>(ins);
    while (ins = ins->next) {
        insreg[ins->addr] = std::make_shared<OpInfo>(ins);
        prev = prev->next = insreg[ins->addr];
    }
    ins = parser.begin;
    while (ins) {
        auto opinfo = insreg[ins->addr];

        if (ins->isJump()) {
            for (uint32_t offset : ins->args) {
                auto it = insreg.find(offset);

                // Invalid jump; jump to next instruction instead
                if (it == insreg.end())
                    offset = ins->next->addr;

                auto target = insreg[offset];
                opinfo->jumpsTo.push_back(target);
                target->jumpsHere.insert(opinfo.get());
            }
        }
        ins = ins->next;
    }
    ins = parser.begin;

    bool modified = false;
    while (ins) {
        auto opinfo = insreg[ins->addr];
        switch (ins->opcode) {
        case OP::pushbyte:
            stack.push(static_cast<double>(static_cast<int8_t>(ins->args[0])));
            break;
        case OP::pushshort:
            stack.push(static_cast<double>(static_cast<int32_t>(ins->args[0])));
            break;
        case OP::pushint:
            stack.push(static_cast<double>(abc->cpool.integers[ins->args[0]]));
            break;
        case OP::pushuint:
            stack.push(static_cast<double>(abc->cpool.uintegers[ins->args[0]]));
            break;
        case OP::pushdouble:
            stack.push(abc->cpool.doubles[ins->args[0]]);
            break;
        case OP::pushtrue:
        case OP::pushfalse:
            stack.push(ins->opcode == OP::pushtrue);
            break;
        case OP::pushstring:
            stack.push(abc->cpool.strings[ins->args[0]]);
            break;
        case OP::dup:
            stack.push(stack.top());
            break;
        case OP::swap: {
            auto tmp = stack.top();
            stack.pop();
            if (!stack.empty())
                std::swap(stack.top(), tmp);
            stack.push(tmp);
            break;
        }
        case OP::OP_not: {
            if (std::holds_alternative<bool>(stack.top())) {
                stack.push(std::monostate());
                break;
            }
            stack.top() = !std::get<bool>(stack.top());
            break;
        }
        case OP::subtract:
        case OP::multiply:
        case OP::divide:
        case OP::add: {
            auto a = pop_value(stack), b = pop_value(stack);
            auto& ops = operations[a.index()];
            if (a.index() != b.index() || a.index() == 0 || ops.find(ins->opcode) == ops.end()) {
                // Not the same type, one is monostate, or the operation is not registered
                // Don't bother trying
                stack.push(std::monostate());
                break;
            }
            modified = true;
            stack.push(ops.at(ins->opcode)(a, b));
            edit_ins(abc, parser, insreg, stack, opinfo, 2);
            break;
        }
        case OP::negate: {
            if (!std::holds_alternative<double>(stack.top())) {
                modified    = true;
                stack.top() = -std::get<double>(stack.top());
                edit_ins(abc, parser, insreg, stack, opinfo, 1);
            }
            break;
        }
        default: {
            if (stack_operations.find(ins->opcode) == stack_operations.end())
                throw std::runtime_error(
                    std::string("Unsupported operation.") + opnames[int(ins->opcode)]);

            if (dynamic_stack_op.find(ins->opcode) != dynamic_stack_op.end()) {
                for (uint32_t i = 0; i < ins->args[dynamic_stack_op.at(ins->opcode)]; ++i)
                    stack.pop();
            }

            auto& [take, put] = stack_operations.at(ins->opcode);
            for (uint32_t i = 0; i < take; ++i)
                stack.pop();

            for (uint32_t i = 0; i < put; ++i)
                stack.push(std::monostate());

            break;
        }
        }
        ins = ins->next;
    }

    if (modified) {
        uint32_t pos = 0;

        // re-compute the instructions position
        ins = parser.begin;
        while (ins) {
            auto opinfo  = insreg[ins->addr];
            opinfo->addr = pos;
            pos += ins->size();
            ins = ins->next;
        }

        ins = parser.begin;
        while (ins) {
            auto opinfo = insreg[ins->addr];

            if (ins->isJump())
                for (auto i = 0; i < ins->args.size(); ++i)
                    ins->args[i] = opinfo->jumpsTo[i]->addr;

            ins = ins->next;
        }

        swf::StreamWriter stream;
        ins = parser.begin;
        while (ins) {
            ins->write(stream);
            ins = ins->next;
        }
        method.code.clear();
        method.code.insert(
            method.code.end(), stream.get_buffer(), stream.get_buffer() + stream.size());
    }
}

File: detfm/src/detfm/simplify.rs
use super::{
    eval::{StackValue, STACK_OPERATIONS},
    jump_info::{JumpInfo, JumpInfoError},
};
use anyhow::Result;
use rabc::{
    abc::{
        constant_pool::PushGetIndex,
        parser::{
            opargs::{PushByteArg, PushDoubleArg, PushShortArg, PushStringArg},
            Instruction, Op, OpCode,
        },
        ConstantPool, Method,
    },
    error::RabcError,
};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum SimplifyError {
    #[error("Cannot pop value from an empty stack.")]
    EmptyStack,
    #[error("Unsupported opcode: {0}")]
    UnsupportedOpCode(OpCode),
    #[error(transparent)]
    JumpInfoError(#[from] JumpInfoError),
    #[error(transparent)]
    RabcError(#[from] RabcError),
}

pub fn simplify_expressions(
    cpool: &mut ConstantPool,
    method: &mut Method,
) -> Result<(), SimplifyError> {
    let instructions = method.parse()?;
    let mut stack: Vec<StackValue> = Vec::with_capacity(method.max_stack as usize);
    let mut modified = false;
    let mut jump_info = JumpInfo::new(method, &instructions);
    for ins in &instructions {
        let mut new_ins = ins.clone();
        match &ins.op {
            #[allow(clippy::cast_possible_wrap)]
            Op::PushByte(arg) => stack.push((arg.value as i8).into()),
            Op::PushShort(arg) => stack.push(arg.value.into()),
            Op::PushInt(arg) => stack.push((*cpool.get_int(arg.value)?).into()),
            Op::PushUint(arg) => stack.push((*cpool.get_uint(arg.value)?).into()),
            Op::PushDouble(arg) => stack.push((*cpool.get_double(arg.value)?).into()),
            Op::PushTrue() => stack.push(true.into()),
            Op::PushFalse() => stack.push(false.into()),
            Op::PushString(arg) => stack.push(cpool.get_str(arg.value)?.clone().into()),
            Op::Dup() => stack.push(stack.last().ok_or(SimplifyError::EmptyStack)?.clone()),
            Op::Swap() => {
                let length = stack.len();
                if length < 2 {
                    return Err(SimplifyError::EmptyStack);
                }
                stack.swap(length - 1, length - 2);
            }
            Op::Not() => match stack.last_mut().ok_or(SimplifyError::EmptyStack)? {
                StackValue::Boolean(value) => *value = !*value,
                value => *value = StackValue::Invalid(),
            },
            Op::Subtract() | Op::Multiply() | Op::Divide() | Op::Add() => {
                let a = stack.pop().ok_or(SimplifyError::EmptyStack)?;
                let b = stack.pop().ok_or(SimplifyError::EmptyStack)?;
                let value = a.op(b, ins.opcode);
                if !matches!(value, StackValue::Invalid()) {
                    modified = true;
                    // remove_ins(&mut new_instructions, &mut new_ins, &mut jump_info, 2);
                    jump_info.pop();
                    jump_info.pop();
                    edit_ins(cpool, &mut new_ins, &value);
                }
                stack.push(value);
            }
            Op::Negate() => {
                if let Some(stack_value) = stack.last_mut() {
                    if let StackValue::Number(value) = stack_value {
                        *value = -*value;
                        modified = true;
                        // remove_ins(&mut new_instructions, &mut new_ins, &mut jump_info, 1);
                        jump_info.pop();
                        edit_ins(cpool, &mut new_ins, stack_value);
                    }
                }
            }
            Op::CallProperty(arg)
                if arg.arg_count == 1
                    && cpool.str(arg.property).is_some_and(|n| n == "Boolean") =>
            {
                let value = stack.pop().ok_or(SimplifyError::EmptyStack)?;
                modified = true;
                // remove_ins(&mut new_instructions, &mut new_ins, &mut jump_info, 2);
                jump_info.pop();
                jump_info.pop();
                edit_ins(cpool, &mut new_ins, &value);
                stack.push(value.to_bool());
            }
            op => {
                let Some([mut take, put]) = STACK_OPERATIONS.get(&ins.opcode) else {
                    return Err(SimplifyError::UnsupportedOpCode(ins.opcode));
                };
                // dynamic stack opcodes
                take += match op {
                    Op::Construct(n) | Op::ApplyType(n) | Op::NewArray(n) => n.arg_count,
                    Op::ConstructProp(n) | Op::CallProperty(n) | Op::CallPropVoid(n) => n.arg_count,
                    _ => 0,
                };

                // consume stack values
                stack.truncate(match stack.len().checked_sub(take as usize) {
                    Some(length) => length,
                    _ => return Err(SimplifyError::EmptyStack),
                });
                // push new stack values
                for _ in 0..*put {
                    stack.push(StackValue::Invalid());
                }
            }
        }
        // new_instructions.push(new_ins);
        jump_info.add(new_ins);
    }

    if modified {
        let new_instructions = jump_info.fix_addresses(method)?;
        method.save_instructions(&new_instructions)?;
    }
    Ok(())
}

/* fn remove_ins(
    instructions: &mut Vec<Instruction>,
    ins: &mut Instruction,
    jump_info: &mut JumpInfo,
    to_remove: usize,
) {
    for _ in 0..to_remove {
        if let Some(before) = instructions.pop() {
            jump_info.remove(before.addr, ins.addr);
            // ins.jumps_here.extend(before.jumps_here);
        }
    }
} */
fn edit_ins(cpool: &mut ConstantPool, ins: &mut Instruction, value: &StackValue) {
    match value {
        StackValue::Number(value) => {
            if value.fract() != 0.0 || value.abs() > 0x8000.into() {
                ins.opcode = OpCode::PushDouble;
                ins.op = Op::PushDouble(PushDoubleArg {
                    value: cpool.doubles.pushi(*value),
                });
            } else if value.abs() > 0x80.into() {
                ins.opcode = OpCode::PushShort;
                ins.op = Op::PushShort(PushShortArg {
                    value: *value as i16,
                });
            } else {
                ins.opcode = OpCode::PushByte;
                ins.op = Op::PushByte(PushByteArg {
                    value: *value as i8 as u8,
                });
            }
        }
        StackValue::String(value) => {
            ins.opcode = OpCode::PushString;
            ins.op = Op::PushString(PushStringArg {
                value: cpool.strings.pushi(value.to_owned()),
            });
        }
        StackValue::Boolean(value) => {
            ins.opcode = if *value {
                OpCode::PushTrue
            } else {
                OpCode::PushFalse
            };
            ins.op = if *value {
                Op::PushTrue()
            } else {
                Op::PushFalse()
            };
        }
        StackValue::Invalid() => {
            panic!("Invalid stack value.")
        }
    }
}
